File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/OfflineToriiException.java
package org.hyperledger.iroha.android.offline;

/** Runtime error raised when offline Torii endpoints reject a request or parsing fails. */
public final class OfflineToriiException extends RuntimeException {

  private static final long serialVersionUID = 1L;
  private final Integer statusCode;
  private final String rejectCode;
  private final String responseBody;

  public OfflineToriiException(final String message) {
    this(message, null, null, null, null);
  }

  public OfflineToriiException(final String message, final Throwable cause) {
    this(message, cause, null, null, null);
  }

  public OfflineToriiException(
      final String message,
      final Integer statusCode,
      final String rejectCode,
      final String responseBody) {
    this(message, null, statusCode, rejectCode, responseBody);
  }

  public OfflineToriiException(
      final String message,
      final Throwable cause,
      final Integer statusCode,
      final String rejectCode,
      final String responseBody) {
    super(message, cause);
    this.statusCode = statusCode;
    this.rejectCode = rejectCode == null || rejectCode.isBlank() ? null : rejectCode;
    this.responseBody = responseBody == null || responseBody.isBlank() ? null : responseBody;
  }

  public java.util.Optional<Integer> statusCode() {
    return java.util.Optional.ofNullable(statusCode);
  }

  public java.util.Optional<String> rejectCode() {
    return java.util.Optional.ofNullable(rejectCode);
  }

  public java.util.Optional<String> responseBody() {
    return java.util.Optional.ofNullable(responseBody);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/OfflineToriiException.kt
package org.hyperledger.iroha.sdk.offline

/** Runtime error raised when offline Torii endpoints reject a request or parsing fails. */
class OfflineToriiException : RuntimeException {

    val statusCode: Int?
    val rejectCode: String?
    val responseBody: String?

    constructor(message: String) : this(message, null, null, null, null)

    constructor(message: String, cause: Throwable?) : this(message, cause, null, null, null)

    constructor(
        message: String,
        statusCode: Int?,
        rejectCode: String?,
        responseBody: String?,
    ) : this(message, null, statusCode, rejectCode, responseBody)

    constructor(
        message: String,
        cause: Throwable?,
        statusCode: Int?,
        rejectCode: String?,
        responseBody: String?,
    ) : super(message, cause) {
        this.statusCode = statusCode
        this.rejectCode = if (rejectCode.isNullOrBlank()) null else rejectCode
        this.responseBody = if (responseBody.isNullOrBlank()) null else responseBody
    }

    companion object {
        private const val serialVersionUID: Long = 1L
    }
}
